trestle/core/rules_validator.py

```diff
@@ -0,0 +1,145 @@
+from trestle.common.list_utils import as_list
+from trestle.core.control_interface import ControlInterface
+from trestle.core.validator import Validator
+from trestle.oscal.component import ComponentDefinition
+from trestle.oscal.ssp import SystemSecurityPlan
+
+logger = logging.getLogger(__name__)
+
+
+class RulesValidator(Validator):
+    """Validator to confirm all rule parameter values are consistent."""
+
+    @staticmethod
+    def val_diff_param_values(existing_param: list, shared_param_values: list) -> list:
+        """
+        Test differences between rule param values.
+
+        args:
+            existing_param: Existing rule param values list.
+            shared_param_values: Current rule param values list.
+
+        returns:
+            Difference between values if exist
+        """
+        for param_value in as_list(existing_param):
+            # order params lists and gets differences
+            param_value['values'].sort()
+            shared_param_values.sort()
+            param_vals_set = set(param_value['values'])
+            difference = [x for x in shared_param_values if x not in param_vals_set]
+            # if there´s a difference in params, log error
+            return difference
+
+    def model_is_valid(self, model: TopLevelOscalModel, quiet: bool) -> bool:
+        """
+        Test if the model is valid.
+
+        args:
+            model: A top level OSCAL model.
+            quiet: Don't report msgs unless invalid.
+
+        returns:
+            True (valid) if the model's rule parameter values are the same accross controls.
+        """
+        # verify if model type is either an SSP of a Component Definition
+        if not isinstance(model, (SystemSecurityPlan, ComponentDefinition)):
+            return True
+        # iterate by each component defined
+        for component in as_list(model.components):
+            shared_params = {}
+            # iterate by each control implementation within the component
+            for control_imp in as_list(component.control_implementations):
+                _, param_dict, _ = ControlInterface.get_rules_and_params_dict_from_item(control_imp)
+                control_imp_rule_param_ids = [d['name'] for d in param_dict.values()]
+                for set_param in as_list(control_imp.set_parameters):
+                    # if not a rule parameter then ignore
+                    if set_param.param_id not in (control_imp_rule_param_ids):
+                        continue
+                    opts_for_parameter = [p['options'] for p in param_dict.values() if p['name'] == set_param.param_id]
+                    shared_params = {
+                        set_param.param_id: [
+                            {
+                                'level': 'ci',
+                                'control_id': '',
+                                'values': set_param.values,
+                                'options': opts_for_parameter
+                            }
```

Comment:
You are keeping a fairly detailed list of information from all the setparameters - but all that matters is the final value that gets set in each control.  So the only thing you really need to build is a dict of `{param: [{control_id: values}]}`  For each rule param there will be a list of controls that reference that param - and have values assigned by a chain of setparams.  All that matters is the final setparam in that chain - that sets the final value.